OSL not working properly when autoshift is enabled

<!-- Provide a general summary of the bug in the Title above -->

<!-- This template is entirely option and can be removed, but is here to help both you and us. -->
<!-- This text and anything on lines wrapped like this one will not show up in the final text. This text is to help us and you. -->

**Describe the bug**

When I have the autoshift enabled (hold long for shift a key) the OSL modifiers are not working as expecting. 
What I expect is to hit the OSL, go to the desired layer, hit one keystroke and come back. However, if the key I hit is not excluded from autoshift then it does not come back to the original layer, I have to hit one of the keys excluded from autoshift (in my case letters) to get out of the layer.

Also, OSL keys started to behave weird, in the sense that I can not hit again to deactivate the layer, i have to hit it several times once it finally decides to come back to the original layer, usually I have to tap it twice in a row and then on the third tap it comes back.
I have the ONESHOT_TAP_TOGGLE set to two.

Regards

**System Information**
 - Keyboard:  ergodox
   - Revision (if applicable):  ez
 - Operating System: windows
 - QMK Firmware version: 0.5.174
 - Any keyboard related software installed? 
  - none

Hello @drashna , thank you for your prompt response.
When I add your code I get this error while compiiling:

```
ycode/process_auto_shift.c: In function 'process_auto_shift':
quantum/process_keycode/process_auto_shift.c:198:9: error: too few arguments to function 'set_oneshot_layer'
         set_oneshot_layer(ONESHOT_PRESSED);
         ^
```

--- a/docs/feature_auto_shift.md
+++ b/docs/feature_auto_shift.md
@@ -51,11 +51,14 @@ By default, Auto Shift is disabled for any key press that is accompanied by one 
 modifiers. Thus, Ctrl+A that you hold for a really long time is not the same
 as Ctrl+Shift+A.
 
-You can re-enable Auto Shift for modifiers by adding another rule to your `rules.mk`
+You can re-enable Auto Shift for modifiers by adding a define to your `config.h`
 
-    AUTO_SHIFT_MODIFIERS = yes
+```c
+#define AUTO_SHIFT_MODIFIERS
+```
 
 In which case, Ctrl+A held past the `AUTO_SHIFT_TIMEOUT` will be sent as Ctrl+Shift+A
+
 
 ## Configuring Auto Shift
 
@@ -65,15 +68,12 @@ behavior of Auto Shift. This is done by setting various variables the
 
 A sample is
 
-    #ifndef CONFIG_USER_H
-    #define CONFIG_USER_H
+```c
+#pragma once
 
-    #include "../../config.h"
-
-    #define AUTO_SHIFT_TIMEOUT 150
-    #define NO_AUTO_SHIFT_SPECIAL
-
-    #endif
+#define AUTO_SHIFT_TIMEOUT 150
+#define NO_AUTO_SHIFT_SPECIAL
+```
 
 ### AUTO_SHIFT_TIMEOUT (Value in ms)
 

--- a/quantum/process_keycode/process_auto_shift.c
+++ b/quantum/process_keycode/process_auto_shift.c
@@ -109,7 +109,9 @@ bool process_auto_shift(uint16_t keycode, keyrecord_t *record) {
         case KC_1 ... KC_0:
 #    endif
 #    ifndef NO_AUTO_SHIFT_SPECIAL
+        case KC_TAB:
         case KC_MINUS ... KC_SLASH:
+        case KC_NONUS_BSLASH:
 #    endif
             autoshift_flush();
             if (!autoshift_enabled) return true;

--- a/quantum/process_keycode/process_auto_shift.c
+++ b/quantum/process_keycode/process_auto_shift.c
@@ -20,19 +20,10 @@
 
 #    include "process_auto_shift.h"
 
-#    define TAP(key)        \
-        register_code(key); \
-        unregister_code(key)
-
-#    define TAP_WITH_MOD(mod, key) \
-        register_code(mod);        \
-        register_code(key);        \
-        unregister_code(key);      \
-        unregister_code(mod)
-
-uint16_t autoshift_time    = 0;
-uint16_t autoshift_timeout = AUTO_SHIFT_TIMEOUT;
-uint16_t autoshift_lastkey = KC_NO;
+static bool autoshift_enabled = true;
+static uint16_t autoshift_time = 0;
+static uint16_t autoshift_timeout = AUTO_SHIFT_TIMEOUT;
+static uint16_t autoshift_lastkey = KC_NO;
 
 void autoshift_timer_report(void) {
     char display[8];
@@ -52,22 +43,16 @@ void autoshift_flush(void) {
         uint16_t elapsed = timer_elapsed(autoshift_time);
 
         if (elapsed > autoshift_timeout) {
-            register_code(KC_LSFT);
+            tap_code16(LSFT(autoshift_lastkey));
+        } else {
+            tap_code(autoshift_lastkey);
         }
 
-        register_code(autoshift_lastkey);
-        unregister_code(autoshift_lastkey);
-
-        if (elapsed > autoshift_timeout) {
-            unregister_code(KC_LSFT);
-        }
-
-        autoshift_time    = 0;
+        autoshift_time = 0;
         autoshift_lastkey = KC_NO;
     }
 }
 
-bool autoshift_enabled = true;
 
 void autoshift_enable(void) { autoshift_enabled = true; }
 void autoshift_disable(void) {
@@ -79,116 +64,72 @@ void autoshift_toggle(void) {
     if (autoshift_enabled) {
         autoshift_enabled = false;
         autoshift_flush();
-    } else {
+    }
+    else {
         autoshift_enabled = true;
     }
 }
 
-bool autoshift_state(void) { return autoshift_enabled; }
+bool get_autoshift_state(void) { return autoshift_enabled; }
+
+uint16_t get_autoshift_timeout(void) { return autoshift_timeout; }
+
+void set_autoshift_timeout(uint16_t timeout) { autoshift_timeout = timeout; }
 
 bool process_auto_shift(uint16_t keycode, keyrecord_t *record) {
-#    ifndef AUTO_SHIFT_MODIFIERS
-    static uint8_t any_mod_pressed;
-#    endif
 
     if (record->event.pressed) {
         switch (keycode) {
-            case KC_ASUP:
-                autoshift_timeout += 5;
-                return false;
+        case KC_ASUP:
+            autoshift_timeout += 5;
+            return true;
 
-            case KC_ASDN:
-                autoshift_timeout -= 5;
-                return false;
+        case KC_ASDN:
+            autoshift_timeout -= 5;
+            return true;
 
-            case KC_ASRP:
-                autoshift_timer_report();
-                return false;
+        case KC_ASRP:
+            autoshift_timer_report();
+            return true;
 
-            case KC_ASTG:
-                autoshift_toggle();
-                return false;
-            case KC_ASON:
-                autoshift_enable();
-                return false;
-            case KC_ASOFF:
-                autoshift_disable();
-                return false;
+        case KC_ASTG:
+            autoshift_toggle();
+            return true;
+        case KC_ASON:
+            autoshift_enable();
+            return true;
+        case KC_ASOFF:
+            autoshift_disable();
+            return true;
 
 #    ifndef NO_AUTO_SHIFT_ALPHA
-            case KC_A:
-            case KC_B:
-            case KC_C:
-            case KC_D:
-            case KC_E:
-            case KC_F:
-            case KC_G:
-            case KC_H:
-            case KC_I:
-            case KC_J:
-            case KC_K:
-            case KC_L:
-            case KC_M:
-            case KC_N:
-            case KC_O:
-            case KC_P:
-            case KC_Q:
-            case KC_R:
-            case KC_S:
-            case KC_T:
-            case KC_U:
-            case KC_V:
-            case KC_W:
-            case KC_X:
-            case KC_Y:
-            case KC_Z:
+        case KC_A ... KC_Z:
 #    endif
 #    ifndef NO_AUTO_SHIFT_NUMERIC
-            case KC_1:
-            case KC_2:
-            case KC_3:
-            case KC_4:
-            case KC_5:
-            case KC_6:
-            case KC_7:
-            case KC_8:
-            case KC_9:
-            case KC_0:
+        case KC_1 ... KC_0:
 #    endif
 #    ifndef NO_AUTO_SHIFT_SPECIAL
-            case KC_MINUS:
-            case KC_EQL:
-            case KC_TAB:
-            case KC_LBRC:
-            case KC_RBRC:
-            case KC_BSLS:
-            case KC_SCLN:
-            case KC_QUOT:
-            case KC_COMM:
-            case KC_DOT:
-            case KC_SLSH:
-            case KC_GRAVE:
-            case KC_NONUS_BSLASH:
-            case KC_NONUS_HASH:
+        case KC_MINUS ... KC_SLASH:
 #    endif
-
-                autoshift_flush();
-                if (!autoshift_enabled) return true;
+            autoshift_flush();
+            if (!autoshift_enabled) return true;
 
 #    ifndef AUTO_SHIFT_MODIFIERS
-                any_mod_pressed = get_mods() & (MOD_BIT(KC_LGUI) | MOD_BIT(KC_RGUI) | MOD_BIT(KC_LALT) | MOD_BIT(KC_RALT) | MOD_BIT(KC_LCTL) | MOD_BIT(KC_RCTL) | MOD_BIT(KC_LSFT) | MOD_BIT(KC_RSFT));
-
-                if (any_mod_pressed) {
-                    return true;
-                }
-#    endif
-
-                autoshift_on(keycode);
-                return false;
-
-            default:
-                autoshift_flush();
+            if (get_mods()) {
                 return true;
+            }
+#    endif
+            autoshift_on(keycode);
+
+            // We need some extra handling here for OSL edge cases
+#    if !defined(NO_ACTION_ONESHOT) && !defined(NO_ACTION_TAPPING)
+            clear_oneshot_layer_state(ONESHOT_OTHER_KEY_PRESSED);
+#    endif
+            return false;
+
+        default:
+            autoshift_flush();
+            return true;
         }
     } else {
         autoshift_flush();

--- a/quantum/process_keycode/process_auto_shift.c
+++ b/quantum/process_keycode/process_auto_shift.c
@@ -20,8 +20,8 @@
 
 #    include "process_auto_shift.h"
 
-static bool autoshift_enabled = true;
-static uint16_t autoshift_time = 0;
+static bool     autoshift_enabled = true;
+static uint16_t autoshift_time    = 0;
 static uint16_t autoshift_timeout = AUTO_SHIFT_TIMEOUT;
 static uint16_t autoshift_lastkey = KC_NO;
 
@@ -48,11 +48,10 @@ void autoshift_flush(void) {
             tap_code(autoshift_lastkey);
         }
 
-        autoshift_time = 0;
+        autoshift_time    = 0;
         autoshift_lastkey = KC_NO;
     }
 }
-
 
 void autoshift_enable(void) { autoshift_enabled = true; }
 void autoshift_disable(void) {
@@ -64,8 +63,7 @@ void autoshift_toggle(void) {
     if (autoshift_enabled) {
         autoshift_enabled = false;
         autoshift_flush();
-    }
-    else {
+    } else {
         autoshift_enabled = true;
     }
 }
@@ -77,61 +75,60 @@ uint16_t get_autoshift_timeout(void) { return autoshift_timeout; }
 void set_autoshift_timeout(uint16_t timeout) { autoshift_timeout = timeout; }
 
 bool process_auto_shift(uint16_t keycode, keyrecord_t *record) {
-
     if (record->event.pressed) {
         switch (keycode) {
-        case KC_ASUP:
-            autoshift_timeout += 5;
-            return true;
+            case KC_ASUP:
+                autoshift_timeout += 5;
+                return true;
 
-        case KC_ASDN:
-            autoshift_timeout -= 5;
-            return true;
+            case KC_ASDN:
+                autoshift_timeout -= 5;
+                return true;
 
-        case KC_ASRP:
-            autoshift_timer_report();
-            return true;
+            case KC_ASRP:
+                autoshift_timer_report();
+                return true;
 
-        case KC_ASTG:
-            autoshift_toggle();
-            return true;
-        case KC_ASON:
-            autoshift_enable();
-            return true;
-        case KC_ASOFF:
-            autoshift_disable();
-            return true;
+            case KC_ASTG:
+                autoshift_toggle();
+                return true;
+            case KC_ASON:
+                autoshift_enable();
+                return true;
+            case KC_ASOFF:
+                autoshift_disable();
+                return true;
 
 #    ifndef NO_AUTO_SHIFT_ALPHA
-        case KC_A ... KC_Z:
+            case KC_A ... KC_Z:
 #    endif
 #    ifndef NO_AUTO_SHIFT_NUMERIC
-        case KC_1 ... KC_0:
+            case KC_1 ... KC_0:
 #    endif
 #    ifndef NO_AUTO_SHIFT_SPECIAL
-        case KC_TAB:
-        case KC_MINUS ... KC_SLASH:
-        case KC_NONUS_BSLASH:
+            case KC_TAB:
+            case KC_MINUS ... KC_SLASH:
+            case KC_NONUS_BSLASH:
 #    endif
-            autoshift_flush();
-            if (!autoshift_enabled) return true;
+                autoshift_flush();
+                if (!autoshift_enabled) return true;
 
 #    ifndef AUTO_SHIFT_MODIFIERS
-            if (get_mods()) {
-                return true;
-            }
+                if (get_mods()) {
+                    return true;
+                }
 #    endif
-            autoshift_on(keycode);
+                autoshift_on(keycode);
 
-            // We need some extra handling here for OSL edge cases
+                // We need some extra handling here for OSL edge cases
 #    if !defined(NO_ACTION_ONESHOT) && !defined(NO_ACTION_TAPPING)
-            clear_oneshot_layer_state(ONESHOT_OTHER_KEY_PRESSED);
+                clear_oneshot_layer_state(ONESHOT_OTHER_KEY_PRESSED);
 #    endif
-            return false;
+                return false;
 
-        default:
-            autoshift_flush();
-            return true;
+            default:
+                autoshift_flush();
+                return true;
         }
     } else {
         autoshift_flush();

--- a/quantum/process_keycode/process_auto_shift.h
+++ b/quantum/process_keycode/process_auto_shift.h
@@ -19,11 +19,7 @@
 #include "quantum.h"
 
 #ifndef AUTO_SHIFT_TIMEOUT
-<<<<<<< HEAD
 #    define AUTO_SHIFT_TIMEOUT 175
-=======
-#   define AUTO_SHIFT_TIMEOUT 175
->>>>>>> Smallish overhaul of Auto-Shift feature
 #endif
 
 bool process_auto_shift(uint16_t keycode, keyrecord_t *record);

--- a/quantum/process_keycode/process_auto_shift.h
+++ b/quantum/process_keycode/process_auto_shift.h
@@ -24,9 +24,9 @@
 
 bool process_auto_shift(uint16_t keycode, keyrecord_t *record);
 
-void autoshift_enable(void);
-void autoshift_disable(void);
-void autoshift_toggle(void);
-bool get_autoshift_state(void);
+void     autoshift_enable(void);
+void     autoshift_disable(void);
+void     autoshift_toggle(void);
+bool     get_autoshift_state(void);
 uint16_t get_autoshift_timeout(void);
-void set_autoshift_timeout(uint16_t timeout);
+void     set_autoshift_timeout(uint16_t timeout);

--- a/quantum/process_keycode/process_auto_shift.h
+++ b/quantum/process_keycode/process_auto_shift.h
@@ -14,13 +14,16 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#ifndef PROCESS_AUTO_SHIFT_H
-#define PROCESS_AUTO_SHIFT_H
+#pragma once
 
 #include "quantum.h"
 
 #ifndef AUTO_SHIFT_TIMEOUT
+<<<<<<< HEAD
 #    define AUTO_SHIFT_TIMEOUT 175
+=======
+#   define AUTO_SHIFT_TIMEOUT 175
+>>>>>>> Smallish overhaul of Auto-Shift feature
 #endif
 
 bool process_auto_shift(uint16_t keycode, keyrecord_t *record);
@@ -28,6 +31,6 @@ bool process_auto_shift(uint16_t keycode, keyrecord_t *record);
 void autoshift_enable(void);
 void autoshift_disable(void);
 void autoshift_toggle(void);
-bool autoshift_state(void);
-
-#endif
+bool get_autoshift_state(void);
+uint16_t get_autoshift_timeout(void);
+void set_autoshift_timeout(uint16_t timeout);

